.github/workflows/main.yml

```diff
@@ -0,0 +1,74 @@
+name: CI
+
+on:
+  push:
+    branches:
+      - master
+  pull_request: {}
+
+jobs:
+  ksc:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          submodules: 'true'
+      - uses: actions/setup-java@v3
+        with:
+          distribution: temurin
+          java-version: 17
+      - name: build-compiler
+        run: |
+          cd compiler
+          export GIT_COMMIT=$(git log -1 --format=%h)
+          export GIT_DATE_ISO=$(TZ=UTC git log -1 --date=iso-strict-local --format=%cd)
+          export GIT_DATE=$(TZ=UTC git log -1 --date=format-local:%Y%m%d.%H%M%S --format=%cd)
+          export KAITAI_STRUCT_VERSION=0.11-SNAPSHOT${GIT_DATE}.${GIT_COMMIT}
+
+          echo "KAITAI_STRUCT_VERSION=$KAITAI_STRUCT_VERSION"
+
+          sbt \
+            compile \
+            compilerJVM/stage \
+            fastOptJS \
+            buildNpmJsFile \
+            buildNpmPackage \
+            compilerJVM/debian:packageBin \
+            compilerJVM/universal:packageBin
+
+# TODO: add compilerJVM/rpm:packageBin
+
+      - name: publish to artifacts
+        uses: actions/upload-artifact@v3
+        with:
+          name: kaitai-struct-compiler
+          path: |
+            compiler/js/target/scala-*/kaitai-struct-compiler-fastopt.js
+            compiler/js/npm
+            compiler/jvm/target/kaitai-struct-compiler_*_all.deb
+            compiler/jvm/target/universal/kaitai-struct-compiler-*.zip
+
+# TODO: add compiler/jvm/target/rpm/RPMS/noarch/kaitai-struct-compiler-*.noarch.rpm
+
+      - name: translator-tests
+        run: |
+          cd tests
+          ./translator-tests || true
+      - name: build-formats
+        run: |
+          cd tests
+          ./build-formats
+      - name: publish formats
+        env:
+          BOT_SSH_KEY: ${{secrets.BOT_SSH_KEY}}
+        run: |
+          ./push_artifacts/git_config_kaitai_bot
+          ./push_artifacts/publish \
+            -o kaitai-io \
+            -r ci_targets \
+            -m "Regen ${GITHUB_REF#refs/heads/*} kaitai-io/kaitai_struct@$GITHUB_SHA" \
+            -- \
+            --exclude=.git \
+            --exclude=.github \
+            --exclude=.travis.yml \
+            tests/compiled
```

Comment:
Please (and we should make sure to have this in all applicable GH Actions workflows [with `on/pull_request`](https://github.com/search?q=org%3Akaitai-io%20pull_request&type=code)):

```diff
       - name: publish formats
         env:
           BOT_SSH_KEY: ${{secrets.BOT_SSH_KEY}}
         run: |
           ./push_artifacts/git_config_kaitai_bot
           ./push_artifacts/publish \
@@ ... @@
             tests/compiled
+        if: github.event_name == 'push' && github.ref == 'refs/heads/master'
```

See also https://kevsoft.net/2020/06/10/running-github-action-steps-and-jobs-only-on-push-to-master.html

In principle, we surely don't want commits like this to be created just because some PR in the `kaitai_struct` repo is updated: https://github.com/kaitai-io/ci_targets/commit/d1655b6f37e9280942a65f90f0fd5916ed259037

To be precise, a contributor who doesn't have write access to this repository wouldn't be able to trigger this publish step, because workflows from forks [don't have access to secrets](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks#about-workflow-runs-from-public-forks), so `BOT_SSH_KEY` would be inaccessible and the publish step would fail. So this supposedly doesn't have security impact, because it happened on this PR just because you opened it from the [`travis_to_gha`](https://github.com/kaitai-io/kaitai_struct/tree/travis_to_gha) branch that is directly in the [kaitai-io / **kaitai_struct**](https://github.com/kaitai-io/kaitai_struct) repository, which is why it did have access to `BOT_SSH_KEY`. But it is undoubtedly unwanted to let the publish step even run in the first place on PRs.

Reply:
@GreyCat This is the unwanted event I had in mind with the above comment:

![image](https://github.com/kaitai-io/kaitai_struct/assets/47499687/3f9e60d7-ffbf-4108-a701-aa5f5fa5c871)

In my view, CI should only be triggered by pushing to `master`, not to other branches.

Reply:
Thanks, let's do it for now.

Eventually I actually wanted to enable PR builds in some way or form. Pushing artifacts into certain artifacts store (and we effectively use git as such) shouldn't be a very privileged operation, and eventually we should support somehow contributors to Kaitai Struct to run their contributions fully tested — by contributing all across compiler, runtimes, tests, etc, and then unifying all that in some kind of common pipeline run.

But I agree that for now, let's focus on just limiting this to what it was before.

Reply:
> Eventually I actually wanted to enable PR builds in some way or form. Pushing artifacts into certain artifacts store (and we effectively use git as such) shouldn't be a very privileged operation, and eventually we should support somehow contributors to Kaitai Struct to run their contributions fully tested — by contributing all across compiler, runtimes, tests, etc, and then unifying all that in some kind of common pipeline run.

Definitely sounds like a good idea, but our infrastructure isn't ready for that. It would be cool to eventually have the ability to switch between the results of different pull requests at https://ci.kaitai.io/. But it also seems to be quite a bit of work.

Reply:
Done, please take a look.